lib/fastlane/plugin/wpmreleasetoolkit/models/build_code.rb

```diff
@@ -0,0 +1,25 @@
+# The `BuildCode` class represents a build code for an app. This could be the Version Code for an Android app or
+# the BUILD_NUMBER xcconfig value used by certain iOS or Mac apps.
+module Fastlane
+  module Models
+    class BuildCode
```

Comment:
I'm wondering if it makes sense to have separate models for `AppVersion` and `BuildCode`, given that generally we derive one from the other.

For example, for an AppVersion of `12.3.4.5`:
 - For iOS
     - The Long Version would be `12.3.4.5`
     - The Short Version would be `12.3` (that's how WPiOS does it at least)
 - For Android
     - Currently some projects have a `versionCode` that is completely uncorrelated from the `versionName` and evolve separately. But this has caused some issues (including when doing hotfixes which complicates the dance by having to intercalate a versionCode before the latest beta…).
     - While other projects, like Tumblr at least, have adopted a model where the versionCode is deduced from the versionName, which means we're not just adding 1 every time but instead create a number that is based on the major, minor, patch and build_number values… allowing us to have those correlated, and also making the case of hotfixes way easier
     - So this means that `versionName` would be `12.3.4.5` (or `12.3.4-rc-5` if you want to keep that format for some projects)
     - But the `versionCode` would be something composed via `format('1%02d%02d%02d%02d', @major, @minor, @patch, @build_number).to_i`, leading to the value `112030405` for `12.3.4.5`. (the leading `1` is to allow us to support more platforms, like we could use a different prefix for mobile app (`1`) vs watch app (`2`) if we need to for example)


If we go this route, that means that we don't need the `BuildCode` model at all anymore. We just need the `AppVersion` model, and then the different `VersionFormatter` subclasses will take care of not only format the `AppVersion` as a `VERSION_LONG` (iOS) / `versionName` (Android) string, but take care of generating a build code (`BUILD_NUMBER` for iOS, `versionCode` for Android) from that same `AppVersion`

Which also means that the caller (the code in our app's `Fastfile`) doesn't have to worry anymore about bumping both and manipulating to model objects and having to remember to keep them in sync (e.g. remembering to increment both the `AppVersion` and the `BuildCode` together in each lanes that bumps those, etc), but instead will only have a single one `AppVersion` model object to manipulate that represent both, which makes sense since they are so intertwined anyway.

Reply:
> the versionCode is deduced from the versionName

This is where all of our apps are going, but for now our projects work differently and can't be quickly updated as part of this project. I'm happy to prefix the class with `_` or something to indicate that it's internal-only and shouldn't be relied on, but IMHO building it in with the intention to deprecate it later is a more pragmatic approach to getting this work shipped soon :)

Reply:
> and also making the case of hotfixes way easier

💯. The `versionCode` dance with Android hotfixes adds a lot of unneeded work.

> format('1%02d%02d%02d%02d', @major, @minor, @patch, @build_number).to_i

I think we may need more planning and discussion before implementing something this. I like the idea, but am slightly weary of making a huge jump in the `versionCode` or `CFBundleVersion` knowing that we can't ever decrease it.

There's also the issue of the DO version numbers using the year, which would make something like `12023190101`, which is larger than the maximum limit for an Android `versionCode`: `2100000000`. We could probably just cut off the `20` from the year (and hope we're still not working at Automattic in 2100, AKA Y2K: Episode 2 😅), but my larger point is that we should discuss it further before we make a decision that can't be reverted. 

> I'm happy to prefix the class with _ or something to indicate that it's internal-only and shouldn't be relied on

TIL that we have [1,853 uppercase characters](https://developer.squareup.com/blog/new-class-naming-rules-in-ruby/) to choose from for starting a class name, but underscore is not one of them 😃 My vote would be to leave the `BuildCode` class as-is for now and revisit the derivative idea at a later point.

Reply:
> This is where all of our apps are going, but for now our projects work differently and can't be quickly updated as part of this project. I'm happy to prefix the class with _ or something to indicate that it's internal-only and shouldn't be relied on, but IMHO building it in with the intention to deprecate it later is a more pragmatic approach to getting this work shipped soon :)

Fair enough.
We'll still need that new version model to support both though, otherwise I won't be able to adopt it for Tumblr for example.

So what about having the `VersionFormatter` classes have a method like `deduce_build_code(from_app_version:)` or similar? That way in apps that have independent `BuildCode` like WP they will still use the `VersionCalculator` to bump both the `AppVersion` and `BuildCode` values each, but for apps that have generated `BuildCode` they'll use `VersionCalculator` to bump `AppVersion` then `VersionFormatter` to generate a `BuildCode` from that new version. Which should cover all our cases?

Reply:
So IMHO `VersionFormatter` definitely should be able to derive the build code – that's part of its job.

My proposal would be a set of actions that assume the build code is part of the version (ie – Tumblr), and a set of actions that assume it's separate. 

The ones where it's part of the version wouldn't use `BuildCode` – they'd just work on a `Version` object.

The ones where we do have to deal with it separately will have their own set of actions to manage them, so for Tumblr they should never be an issue.

> There's also the issue of the DO version numbers using the year, which would make something like 12023190101, which is larger than the maximum limit for an Android versionCode: 2100000000. We could probably just cut off the 20 from the year

That's an option, though my suggestion for build code would be to do something like ceil(n / 1000) + (n % 1000) – we'll never overflow, but it should always work

Reply:
I added a method to `build_code_calculator` for creating the derived build code: https://github.com/wordpress-mobile/release-toolkit/commit/9550fc50a874933f811d6c92f106c0257e69514a

I decided to put it into the `build_code_calculator` class because it is a value that's being calculated (**next**_build_code) just like the values that are calculated in the `version_calculator` classes (**next**_release_version, etc.) Based on that, it seemed to me like it would be better placed in the Build Code class instead of the Version Calculator class. 

@AliSoftware does that work for what Tumblr needs? I believe it does based on what I saw in Tumblr's `version_helper.rb`, but please let me know if it needs to be updated.